### PR TITLE
added support for platformio

### PR DESCRIPTION
**Pull request by pinkit-cwa**
CovidBracelet was moved to platformio.

PlatformIO is a framework that supports a lot of embedded development platforms, inluding esp-idf as well as zephyr.

One drawback is that PlaformIO is using esp-idf v4.0.1 instead of v4.1. That is why I had to downgrade wifi.c to esp-idf v4.0 API. I am not sure if my implementation is correct - cannot test because I do not yet have a suitable board.

I implemented this switch by converting wifi.c into a directory with various implementations - in this case two different esp-idf versions. I intend to do the same for bluetooth support for esp-idf vs. zephyr.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
+.pio
```
Comment (pinkit-cwa) [inline .gitignore:1]: .pio is PlatformIO's build directory

File: .gitignore [hunk 2]
```diff
@@ -43,4 +44,4 @@ dependencies.lock
 sdkconfig
 sdkconfig.old
 
-.local
\ No newline at end of file
+.local
```
Comment (pinkit-cwa) [inline .gitignore:47]: added line feed at EOF

File: components/wifi-controller/CMakeLists.txt [hunk 3]
```diff
@@ -1,8 +1,8 @@
...
     SRCS 
-        "wifi-controller.c"
+        "espidf-v4.1.c"
     INCLUDE_DIRS "."
     PRIV_REQUIRES
         esp_wifi
         nvs_flash
-)
\ No newline at end of file
+)
```
Comment (pinkit-cwa) [inline components/wifi-controller/CMakeLists.txt:8]: added line feed at EOF

File: components/wifi-controller/CMakeLists.txt [hunk 4]
```diff
@@ -1,8 +1,8 @@
 idf_component_register(
     SRCS 
-        "wifi-controller.c"
+        "espidf-v4.1.c"
```
Comment (pinkit-cwa) [inline components/wifi-controller/CMakeLists.txt:3]: created two versions of wifi-controller.c: one using API version 4.0 and one using API version 4.1.
4.1 is identical to wifi-controller.c and used for "latest" esp-idf.
4.0 is used for PlatformIO, as PlatformIOuses esp-idf 4.0.1

File: components/wifi-controller/espidf-v4.0.c [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+wifi_ap_record_t *wifi_controller_connection(void)
+{
+    if (esp_wifi_sta_get_ap_info(&current_wifi_ap) == ESP_OK)
+    {
+        ESP_LOGD(WIFI_LOG, "Current AP: %s", current_wifi_ap.ssid);
+        return &current_wifi_ap;
+    }
+    return NULL;
+}
```
Comment (pinkit-cwa) [inline components/wifi-controller/espidf-v4.0.c:176]: Please compare this files contents with espidf-v4.1.c, and confirm that there are no unintentional changes.
Intention was to build a version for API version 4.0.

File: components/wifi-controller/espidf-v4.1.c [hunk 6]
```diff
@@ -182,4 +182,4 @@ wifi_ap_record_t *wifi_controller_connection(void)
         return &current_wifi_ap;
     }
     return NULL;
-}
\ No newline at end of file
+}
```
Comment (pinkit-cwa) [inline components/wifi-controller/espidf-v4.1.c:185]: added line feed at EOF

File: include/README [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+header file names, and at most one dot.
+
+Read more about using header files in official GCC documentation:
+
+* Include Syntax
+* Include Operation
+* Once-Only Headers
+* Computed Includes
+
+https://gcc.gnu.org/onlinedocs/cpp/Header-Files.html
```
Comment (pinkit-cwa) [inline include/README:39]: automaticallly generated by "platformio init"

File: include/mbedtls+HKDF.h [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+ * Caller:
+ *
+ * Requires: MBEDTLS_MD_C
+ *
+ * This module adds support for the Hashed Message Authentication Code
+ * (HMAC)-based key derivation function (HKDF).
+ */
+#ifndef MBEDTLS_HKDF_C
+#define MBEDTLS_HKDF_C
+#endif
```
Comment (pinkit-cwa) [inline include/mbedtls+HKDF.h:16]: Used to turn on HKDF in esp-idf 4.0 contained mbedTLS.

This is included at the end of mbedtls/esp_config.h via
`-DMBEDTLS_USER_CONFIG_FILE=\"mbedtls+HKDF.h\"`
. See platformio.ini .

File: lib/README [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+  ...
+}
+
+```
+
+PlatformIO Library Dependency Finder will find automatically dependent
+libraries scanning project source files.
+
+More information about PlatformIO Library Dependency Finder
+- https://docs.platformio.org/page/librarymanager/ldf.html
```
Comment (pinkit-cwa) [inline lib/README:46]: automaticallly generated by "platformio init"

File: platformio.ini [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+;   Build options: build flags, source filter
+;   Upload options: custom upload port, speed and extra flags
+;   Library options: dependencies, extra library storages
+;   Advanced options: extra scripting
+;
+; Please visit documentation for the other options and examples
+; https://docs.platformio.org/page/projectconf.html
+
+[platformio]
+src_dir = main
```
Comment (pinkit-cwa) [inline platformio.ini:12]: platformio default source directory is "src". This is used to adjust to esp-idf preference of "main".

File: platformio.ini [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+[env]
+build_flags = 
+    -Iinclude/tls_config
+    -DFRAMEWORK_$PIOFRAMEWORK
```
Comment (pinkit-cwa) [inline platformio.ini:17]: [env] section contains defaults for all platforms; must still be explicitly referenced using ${env. ... }, see below.

File: platformio.ini [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+platform = espressif32
+framework = espidf
+board = esp32dev
+build_flags = 
+    ${env.build_flags}
+    -Iinclude
+    -DMBEDTLS_USER_CONFIG_FILE=\"mbedtls+HKDF.h\"
+board_build.embed_files =
+    components/ena/test/export.bin
+    components/ena-cwa/certs/telekom.pem
```
Comment (pinkit-cwa) [inline platformio.ini:29]: CMakeLists.txt are parsed differently than in esp-idf. This is compensated here.

-Iinclude  -DMBEDTLS_USER_CONFIG_FILE=\"mbedtls+HKDF.h\"
is used to enable HKDF.

board_build.embed_files is PlatformIOs way for including verbatim file contents into build result.

File: platformio.ini [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+board = ng_beacon ; tbd
+framework = zephyr
+monitor_speed = 115200
+test_ignore = test_desktop
+build_flags = 
+    ${env.build_flags}
+    # For testing: -DUNITY_EXCLUDE_SETJMP_H=1
+    -DEN_INCLUDE_ZEPHYR_DEPS=1
+    -DEN_INIT_MBEDTLS_ENTROPY=0
+lib_deps = exposure-notification
```
Comment (pinkit-cwa) [inline platformio.ini:43]: does not work yet, and probably will be hard to implement. This platform is a 15EUR nRF51822 based over-the-air-flashable fitness tracker bracelet.
Just 32kBytes of RAM and 256kBytes of flash. Might be enough for a "tracking only" device for a few days, with off-loading to and risk calculation on smartphone.

File: platformio.ini [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+board = nrf52840_dk
+framework = zephyr
+monitor_speed = 115200
+test_ignore = test_desktop
+build_flags = 
+    ${env.build_flags}
+    # For testing: -DUNITY_EXCLUDE_SETJMP_H=1
+    -DEN_INCLUDE_ZEPHYR_DEPS=1
+    -DEN_INIT_MBEDTLS_ENTROPY=0
+lib_deps = exposure-notification
```
Comment (pinkit-cwa) [inline platformio.ini:56]: This is the platform used for CovidBracelet

File: platformio.ini [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+test_ignore = test_embedded
+lib_compat_mode = off
+lib_deps = mbedtls@~2
+build_flags = 
+    ${env.build_flags}
+    -Iinclude/desktop
+    -Wno-nullability-completeness
+    -DMBEDTLS_USER_CONFIG_FILE='"user-tls.conf"'
+src_filter = --<.src/>   
+targets = test
```
Comment (pinkit-cwa) [inline platformio.ini:70]: This is for unit-tests on the local machine.

File: test/README [hunk 16]
```diff
@@ -0,0 +1,11 @@
...
+This directory is intended for PIO Unit Testing and project tests.
+
+Unit Testing is a software testing method by which individual units of
+source code, sets of one or more MCU program modules together with associated
+control data, usage procedures, and operating procedures, are tested to
+determine whether they are fit for use. Unit testing finds problems early
+in the development cycle.
+
+More information about PIO Unit Testing:
+- https://docs.platformio.org/page/plus/unit-testing.html
```
Comment (pinkit-cwa) [inline test/README:11]: automaticallly generated by "platformio init"

File: zephyr/CMakeLists.txt [hunk 17]
```diff
@@ -0,0 +1,6 @@
+cmake_minimum_required(VERSION 3.13.1)
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
+project(esp-ena)
+
+FILE(GLOB app_sources ../src/*.c*)
+target_sources(app PRIVATE ${app_sources})
```
Comment (pinkit-cwa) [inline zephyr/CMakeLists.txt:6]: not usable yet. copied from CovdiBracelet, and needs to be fixed (use main/main.c instead of src/*)

**Comment by Lurkars** [comment]
merged into new branch 'zephyr'

**Pull request closed without merging**